pkg/endpoint/policy.go

```diff
@@ -429,24 +430,39 @@ func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMe
 			e.getLogger().Debug("Dequeued endpoint from build queue")
 
 			regenContext.DoneFunc = doneFunc
-			err := e.regenerate(owner, regenContext)
-			doneFunc() // in case not called already
 
-			repr, reprerr := monitorAPI.EndpointRegenRepr(e, err)
-			if reprerr != nil {
-				e.getLogger().WithError(reprerr).Warn("Notifying monitor about endpoint regeneration failed")
-			}
-
-			if err != nil {
-				buildSuccess = false
-				if reprerr == nil && !option.Config.DryMode {
-					owner.SendNotification(monitorAPI.AgentNotifyEndpointRegenerateFail, repr)
-				}
-			} else {
-				buildSuccess = true
-				if reprerr == nil && !option.Config.DryMode {
-					owner.SendNotification(monitorAPI.AgentNotifyEndpointRegenerateSuccess, repr)
+			epEvent := eventqueue.NewEvent(&EndpointRegenerationEvent{
+				owner:        owner,
+				regenContext: regenContext,
+				ep:           e,
+			})
+
+			e.QueueEvent(epEvent)
+
+			select {
+			case result, ok := <-epEvent.EventResults:
+				var regenError error
+				if ok {
+					// Regeneration is done, allow other builds to occur.
+					doneFunc()
+					log.Warningf("EV TYPE FROM REGEN: %T", reflect.TypeOf(result))
+					regenResult := result.(*EndpointRegenerationResult)
+					regenError = regenResult.err
+
+					// Build was successful whether regeneration errored out of not.
+					buildSuccess = regenError == nil
+				} else {
+					// This may be unnecessary(?) since 'closing' of the results
+					// channel means that event has been cancelled?
+					e.getLogger().Debug("regeneration was cancelled")
+					doneFunc()
 				}
+				// Notify monitor of result of regeneration.
+				e.notifyEndpointRegeneration(owner, regenError)
+			case <-epEvent.Cancelled:
```

Comment:
Consider factoring out handling of 'Canceled' and 'Result' events into separate functions which are generic for a given `EventHandler`. Not sure how clear this would be, though. Each could be the following:

* `CancelFunc func()`
* `ResultFunc func()`